### ISSUE TITLE
CA1859 should understand boxing and IEnumerable

### ISSUE THREAD

**Issue by geeknoid** [dotnet/roslyn-analyzers#7114]
### Analyzer

**Diagnostic ID**: [CA1859](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1859)

### Describe the improvement

I just reviewed some code that did this:

```csharp
  StringValues value = ...
  Foo(values);
  ...
  private void Foo(IEnumerable<string> values)
  {
      foreach (var v in values) { ... }
  }
```

So in the above, there is boxing happening on the call to Foo, and now because of the foreach there is an allocation for the enumerator, and there are two interface calls performed everytime through the loop (MoveNext and Current on the iterator). Switching the parameter to be of type StringValues fixes all this.

* As a general pattern, CA1859 should suggest eliminating boxing when possible. If a method takes an interface type as parameter and all call sites are passing the same struct, then CA1859 should recommend to change the parameter type to be the struct type. The analyzer doesn't currently consider boxing in its analysis but it definitely could. A similar thing can happen if the return value of a method is being boxed and all returns in the method produce the same struct type.

* As a specialization, CA1859 could gain explicit knowledge of the fancy logic behind foreach in the compiler. If you have a method that takes an IEnumerable<T> but all call sites are passing in a List<T>, the rule will often not recommend upgrading the parameter to a List<T> because it will not recognize the fact the "foreach" loop done in the method is inducing interface dispatch, which would go away if the List<T> was accepted as a parameter. So "if calling an IEnumerable and you have a type which has a struct Enumerator, then recommend upgrading to that type".

* My team has a separate rule which says "if you have an IList or IReadOnlyList in hand, please use a 'for' loop instead of 'foreach' loop since the 'for' loop is 2x faster and doesn't allocate". To help this optimization to be possible, CA1859 can flag cases where IENumerable can be upgraded to IList/IReadOnlyList.
